disnake/shard.py

```diff
@@ -426,28 +427,36 @@ async def launch_shard(self, gateway: str, shard_id: int, *, initial: bool = Fal
         self.__shards[shard_id] = ret = Shard(ws, self, self.__queue.put_nowait)
         ret.launch()
 
-    async def launch_shards(self) -> None:
+    async def launch_shards(self, *, ignore_session_start_limit: bool = False) -> None:
         shard_count, gateway, session_start_limit = await self.http.get_bot_gateway()
-        if self.shard_count is None:
-            self.shard_count = shard_count
 
         self.session_start_limit = SessionStartLimit(session_start_limit)
 
+        if self.shard_count is None:
+            self.shard_count = shard_count
+
         self._connection.shard_count = self.shard_count
 
         shard_ids = self.shard_ids or range(self.shard_count)
         self._connection.shard_ids = shard_ids
 
+        if not ignore_session_start_limit and self.session_start_limit.remaining < self.shard_count:
```

Comment:
```suggestion
        if not ignore_session_start_limit and self.session_start_limit.remaining < len(shard_ids):
```
Not necessarily all shards will be launched by this instance

Reply:
Yes, and that's why its like this, as not all shards will be launched by this instance, but we need to think about all instances as a whole.

Ex: 15 counts remain, and 30 shards want to be launched. If there's three instances of 10 shards each, the token will be reset because all three instances had enough room for themselves, but not when working together.

Reply:
That assumes that all instances first fetch the `/gateway/bot` endpoint, and only once all of them have the data, they start launching shards. If they instead start sequentially, i.e. instance 1 fetches gateway + launches shards, then instance 2 fetches gateway + launches shards, ..., some of the later instances might refuse to start.

As an example, 40 remain and we want to launch 3 instances with 10 shards each. With the current implementation:
- instance 1 fetches, sees 40 remaining, checks that there are 30 left, succeeds
- instance 2 fetches, sees 30 remaining, checks that there are 30 left, succeeds
- instance 3 fetches, sees 20 remaining, fails because it wants the limit to be `>=30`, even though it'd only launch the remaining 10 shards

There is no solution to this that doesn't involve some degree of coordination between instances, honestly not sure which one's the better option here:
- if `self.shard_count` is used and the instances start one after the other, you might get a `SessionStartLimitReached` error even though it would've been fine to continue
- if `len(shard_ids)` is used and all instances fetch the gateway in parallel, there's a risk of exceeding the limit

Reply:
To be fair, the latter only occurs when someone has multiple processes (as there's no advantage to multiple clients in a single process), which means that in the latter, they can build their own protection for this and their own clustering system. 

But when having an autoshardedclient, there's one client for all of the shards, and only one shard_count for all of the shards in total. So when using the library as its written, the shards are all managed in a single process. Once a bot is larger and writes its own sharding manager, they can use this as inspiration to write their own identify limit reached protection. That's also why we allow disabling this check.